cdap-security/src/main/java/co/cask/cdap/security/tools/GeneratedCertKeyStores.java

```
@@ -0,0 +1,155 @@
+   * @return A X.509 certificate
+   */
+  private static X509Certificate getCertificate(String dn, KeyPair pair, int days, String algorithm) {
+    // Calculate the validity interval of the certificate
+    Date from = new Date();
+    Date to = DateUtils.addDays(from, days);
+    CertificateValidity interval = new CertificateValidity(from, to);
+    // Generate a random number to use as the serial number for the certificate
+    BigInteger sn = new BigInteger(64, new SecureRandom());
+    // Create the name of the owner based on the provided distinguished name
+    try {
+      X500Name owner = new X500Name(dn);
+      // Create an info objects with the provided information, which will be used to create the certificate
+      X509CertInfo info = new X509CertInfo();
+      info.set(X509CertInfo.VALIDITY, interval);
+      info.set(X509CertInfo.SERIAL_NUMBER, new CertificateSerialNumber(sn));
+      // This certificate will be self signed, hence the subject and the issuer are same.
+      info.set(X509CertInfo.SUBJECT, new CertificateSubjectName(owner));
+      info.set(X509CertInfo.ISSUER, new CertificateIssuerName(owner));
+      info.set(X509CertInfo.KEY, new CertificateX509Key(pair.getPublic()));
+      info.set(X509CertInfo.VERSION, new CertificateVersion(CertificateVersion.V3));
+      AlgorithmId algo = new AlgorithmId(AlgorithmId.md5WithRSAEncryption_oid);
+      info.set(X509CertInfo.ALGORITHM_ID, new CertificateAlgorithmId(algo));
+      // Create the certificate and sign it with the private key
+      X509CertImpl cert = new X509CertImpl(info);
+      PrivateKey privateKey = pair.getPrivate();
+      cert.sign(privateKey, algorithm);
+      return cert;
+    } catch (CertificateException | IOException | NoSuchAlgorithmException | SignatureException
+      | InvalidKeyException | NoSuchProviderException e) {
+      throw new RuntimeException("SSL is enabled but an certificate could not be generated. A certificate is required" +
```

Since this is a private method, it's better to propagate the exception and let caller to handle it, otherwise the exception will get double wrapped.